### PR TITLE
Adding Scala 2.13 support part 6: Major code changes

### PR DESCRIPTION
**Pull request by davidrabinowitz**
Co-authored-by: Vinay Londhe <vinaylondhe@google.com>

### CODE REVIEW

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/BigQueryRDDFactory.java [hunk 1]
```diff
@@ -151,6 +156,60 @@ public RDD<?> createRddFromTable(
...
+  @VisibleForTesting
+  RDD<InternalRow> createRDD(
+      SQLContext sqlContext,
+      Partition[] partitions,
+      ReadSession readSession,
+      Schema bqSchema,
+      String[] columnsInOrder,
+      SparkBigQueryConfig options,
+      BigQueryClientFactory bigQueryClientFactory) {
+    // Unfortunately we need to use reflection here due to a cyclic dependency issue, and the fact
```
Comment (VinayLondhe14) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/BigQueryRDDFactory.java:169]: I understand that the RDD constructor dependencies are different between Scala 2.12 and 2.13. But, how does the dependency between BigQueryRDD and BigQueryPartition matter here?

File: spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.spark.sql.catalyst.expressions.NamedExpression;
+import org.apache.spark.sql.types.StructType;
+import scala.collection.JavaConverters;
+import scala.collection.mutable.ListBuffer;
+
+public class Scala213SparkSqlUtils extends SparkSqlUtils {
+
+  @Override
+  public boolean supportsScalaVersion(String scalaVersion) {
+    return scalaVersion.compareTo("2.13") > 0;
```
Comment (VinayLondhe14) [inline spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java:36]: Should this be >=0?

File: spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  @Override
+  public InternalRow rowToInternalRow(Row row) {
+    return InternalRow.fromSeq(row.toSeq());
+  }
+
+  @Override
+  public ExpressionEncoder<Row> createExpressionEncoder(StructType schema) {
```
Comment (VinayLondhe14) [inline spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java:45]: Can we add a comment here on why we need this method? Something similar to the comment added for the `toAttributes` method below

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDDTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+package com.google.cloud.spark.bigquery.direct;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.junit.Test;
+
+// Ported this class from Scala to Java with no change in functionality
```
Comment (VinayLondhe14) [inline spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDDTest.java:25]: Nit: remove comment

File: spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support/pom.xml [hunk 5]
```diff
@@ -19,10 +19,15 @@
...
+    <dependency>
+      <groupId>${project.groupId}</groupId>
+      <artifactId>spark-bigquery-connector-common</artifactId>
+      <version>${project.version}</version>
+    </dependency>
     <dependency>
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-sql_2.12</artifactId>
-      <version>3.0.0</version>
+      <version>3.2.0</version>
```
Comment (VinayLondhe14) [inline spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support/pom.xml:30]: Should this be 3.3.0?

File: spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.13/pom.xml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                            <sources>
+                                <source>src/test_2.13/java</source>
+                            </sources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (VinayLondhe14) [inline spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.13/pom.xml:53]: Nit: add extra line at the end

File: spark-bigquery-scala-212-support/src/main/java/com/google/cloud/spark/bigquery/direct/PreScala213BigQueryRDD.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.spark.Partition;
+import org.apache.spark.SparkContext;
+import org.apache.spark.TaskContext;
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.sql.catalyst.InternalRow;
+import scala.collection.Seq;
+import scala.collection.Seq$;
+
+// Ported this class from Scala to Java with no change in functionality
+class PreScala213BigQueryRDD extends RDD<InternalRow> {
```
Comment (VinayLondhe14) [inline spark-bigquery-scala-212-support/src/main/java/com/google/cloud/spark/bigquery/direct/PreScala213BigQueryRDD.java:42]: Can you add a comment on why we need this class? Similar to the one added for reflection in BigQueryRDDFactory

File: spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/org/apache/spark/sql/Scala213SparkSqlUtils.java:36]: In practice the version is always 2.13.x, which is "larger" than 2.13

File: pom.xml [hunk 8]
```diff
@@ -86,8 +87,9 @@
         <activeByDefault>false</activeByDefault>
       </activation>
       <modules>
-        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support</module>
         <module>spark-bigquery-dsv1/spark-bigquery-dsv1-parent</module>
+        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-spark2-support</module>
+        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support</module>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply pom.xml:92]: That's true, but at the moment the factory has an import dependency on both implementation (no service loader is used). Therefore it must be in the classpath In any case, both sparkN-support has just 1 class in them

File: pom.xml [hunk 9]
```diff
@@ -86,8 +87,9 @@
         <activeByDefault>false</activeByDefault>
       </activation>
       <modules>
-        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support</module>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply pom.xml:89]: I've added all spark-bigquery-pushdown to the dsv1 profile

File: pom.xml [hunk 10]
```diff
@@ -117,6 +120,26 @@
...
       </modules>
     </profile>
+    <profile>
+      <id>dsv1_2.13</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
+      <modules>
+        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-parent</module>
+        <module>spark-bigquery-dsv1/spark-bigquery-dsv1-spark2-support</module>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply pom.xml:130]: Same answer - yes to to class loading issue

File: spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support/pom.xml (hunk 5 above)
Comment (davidrabinowitz) [reply spark-bigquery-dsv1/spark-bigquery-dsv1-spark3-support/pom.xml:30]: I prefer to use the lowest spark version that I can

**Review by VinayLondhe14 — APPROVED** [review]

**Pull request merged**